machine-learning/sagemaker-edge-manager/artifacts/com.greengrass.SageMakerEdgeManager.ImageClassification/1.0.0/image_classification/inference.py

```diff
@@ -0,0 +1,126 @@
+from os import path
+from threading import Thread, Timer
+from time import sleep
+
+import config_utils
+import IPCUtils as ipc_utils
+from agent_pb2_grpc import AgentStub
+from grpc import insecure_channel
+from prediction_utils import load_image, load_model, predict_from_image, unload_model
+
+
+def set_configuration(config):
+    r"""
+    Sets a new config object with the combination of updated and default configuration as applicable.
+    Calls inference code with the new config and indicates that the configuration changed.
+    """
+    new_config = {}
+
+    if "ImageName" in config:
+        new_config["image_name"] = config["ImageName"]
+    else:
+        new_config["image_name"] = config_utils.DEFAULT_IMAGE_NAME
+        config_utils.logger.warning(
+            "Using default image name: {}".format(config_utils.DEFAULT_IMAGE_NAME)
+        )
+
+    if "ImageDirectory" in config:
+        new_config["image_dir"] = config["ImageDirectory"]
+    else:
+        new_config["image_dir"] = config_utils.IMAGE_DIR
+        config_utils.logger.warning(
+            "Using default image directory: {}".format(config_utils.IMAGE_DIR)
+        )
+
+    if "InferenceInterval" in config:
+        new_config["prediction_interval_secs"] = config["InferenceInterval"]
+    else:
+        new_config["prediction_interval_secs"] = config_utils.DEFAULT_PREDICTION_INTERVAL_SECS
+        config_utils.logger.warning(
+            "Using default inference interval: {}".format(
+                config_utils.DEFAULT_PREDICTION_INTERVAL_SECS
+            )
+        )
+
+    if "PublishResultsOnTopic" in config:
+        config_utils.TOPIC = config["PublishResultsOnTopic"]
+    else:
+        config_utils.TOPIC = ""
+        config_utils.logger.warning("Topic to publish inference results is empty.")
+
+    if config_utils.edge_agent_socket_change:
+        config_utils.edge_agent_socket_change = False
+        if "UnixSocketName" in config:
+            while not path.exists(config["UnixSocketName"]):
+                config_utils.logger.info(
+                    "Configuration changed. Waiting for the edge agent to create the socket..."
+                )
+                sleep(1)
+            config_utils.agent_client = AgentStub(
+                insecure_channel(
+                    "unix://{}".format(config["UnixSocketName"]),
+                    options=(("grpc.enable_http_proxy", 0),),
+                )
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *insecure\_channel*. The resource is allocated by call *grpc.insecure_channel*. Currently, there are execution paths that do not contain closure statements, for example, when *agent\_pb2\_grpc.AgentStub()* throws an exception. To prevent this resource leak,  assign the object returned by *insecure\_channel()* to a local variable and close it in a try-finally block.